sled-agent/src/services.rs

```diff
@@ -322,188 +376,223 @@ impl ServiceManager {
     // prefix is spawned onto a background task.
     async fn advertise_prefix_of_address(&self, address: Ipv6Addr) {
         let subnet = Ipv6Subnet::new(address);
-        if self.advertised_prefixes.lock().await.insert(subnet) {
-            self.ddmd_client.advertise_prefix(subnet);
+        if self.inner.advertised_prefixes.lock().await.insert(subnet) {
+            self.inner.ddmd_client.advertise_prefix(subnet);
         }
     }
 
-    // Populates `existing_zones` according to the requests in `services`.
-    //
-    // At the point this function is invoked, IP addresses have already been
-    // allocated (by either RSS or Nexus). However, this function explicitly
-    // assigns such addresses to interfaces within zones.
-    async fn initialize_services_locked(
+    // Check the services intended to run in the zone to determine whether any
+    // physical devices need to be mapped into the zone when it is created.
+    fn devices_needed(
         &self,
-        existing_zones: &mut Vec<RunningZone>,
-        requests: &Vec<ServiceZoneRequest>,
-    ) -> Result<(), Error> {
-        // TODO(https://github.com/oxidecomputer/omicron/issues/726):
-        // As long as we ensure the requests don't overlap, we could
-        // parallelize this request.
-        for req in requests {
-            info!(
-                self.log,
-                "Ensuring service zone is initialized: {:?}", req.zone_name
-            );
-            // Before we bother allocating anything for this request, check if
-            // this service has already been created.
-            let expected_zone_name =
-                InstalledZone::get_zone_name(&req.zone_name, None);
-            if existing_zones.iter().any(|z| z.name() == expected_zone_name) {
-                info!(
-                    self.log,
-                    "Service zone {} already exists", req.zone_name
-                );
-                continue;
-            } else {
-                info!(
-                    self.log,
-                    "Service zone {} does not yet exist", req.zone_name
-                );
+        req: &ServiceZoneRequest,
+    ) -> Result<Vec<String>, Error> {
+        let mut devices = vec![];
+        for svc in &req.services {
+            match svc {
```

Comment:
I took a slightly different approach than https://github.com/oxidecomputer/omicron/pull/1820 here - I didn't want to introduce the `ServiceVariant`, in *addition* to the `ServiceType`, so I just directly matched on the variants here.

Reply:
So the only thing we lose is the generic "get_service()" utility function?  That seems reasonable.

Reply:
Yup, basically.